builtin/logical/ssh/path_issue_sign.go

```diff
@@ -0,0 +1,539 @@
+package ssh
+
+import (
+	"context"
+	"crypto/dsa"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-secure-stdlib/parseutil"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/certutil"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	"golang.org/x/crypto/ssh"
+)
+
+var ecCurveBitsToAlgoName = map[int]string{
+	256: ssh.KeyAlgoECDSA256,
+	384: ssh.KeyAlgoECDSA384,
+	521: ssh.KeyAlgoECDSA521,
+}
+
+// If the algorithm is not found, it could be that we have a curve
+// that we haven't added a constant for yet. But they could allow it
+// (assuming x/crypto/ssh can parse it) via setting a ec: <keyBits>
+// mapping rather than using a named SSH key type, so erring out here
+// isn't advisable.
+
+type creationBundle struct {
+	KeyID           string
+	ValidPrincipals []string
+	PublicKey       ssh.PublicKey
+	CertificateType uint32
+	TTL             time.Duration
+	Signer          ssh.Signer
+	Role            *sshRole
+	CriticalOptions map[string]string
+	Extensions      map[string]string
+}
+
+func (b *backend) pathSignIssueCertificateHelper(ctx context.Context, req *logical.Request, data *framework.FieldData, role *sshRole, publicKey ssh.PublicKey) (*logical.Response, error) {
+	// Note that these various functions always return "user errors" so we pass
+	// them as 4xx values
+	keyID, err := b.calculateKeyID(data, req, role, publicKey)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), err
```

Comment:
Not to nitpick your nitpick, but this was just a copy/paste refactor of the existing code:

https://github.com/hashicorp/vault/blob/7c72d9d3d6f127ff368d6f47634684c6fb60cb22/builtin/logical/ssh/path_sign.go#L125-L128

And worst, it seems consistently used that way. :/

Reply:
Actually there is a behavior change here, as the callers are wrapping any error into a logical.ErrorResponse so some existing error scenarios we did return a 500 will be switched to a 400.

Reply:
Also makes sense, I remember making this change during development because was having some kind of error. The latest commit reverts these returns to how they were initially (also not consistent). If consistency in the return statements is what we are looking for, I can make the change to `return nil, err` as suggested.

Reply:
Thanks! What you have changed it too, reverting to what it was, is the correct approach, much appreciated.